File: metacasa-ios/MetaCasa/Models/Installment.swift
import Foundation

/// Plan de cuotas (ej. iPhone 15 en 12 cuotas).
/// Port de `CuotaForm` / `CuotaCard` del web (App.jsx:1221-1555).
struct InstallmentPlan: Codable, Identifiable, Hashable, Sendable {
    let id: UUID
    var householdId: UUID
    var name: String
    var totalAmount: Decimal
    var totalInstallments: Int
    var currency: String
    var startYear: Int
    var startMonth: Int
    var category: String?
    var accountId: UUID?
    var note: String?
    var status: PlanStatus
    let createdBy: UUID
    let createdAt: Date?
    var updatedAt: Date?

    enum CodingKeys: String, CodingKey {
        case id
        case householdId = "household_id"
        case name
        case totalAmount = "total_amount"
        case totalInstallments = "total_installments"
        case currency
        case startYear = "start_year"
        case startMonth = "start_month"
        case category
        case accountId = "account_id"
        case note
        case status
        case createdBy = "created_by"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }

    /// Cuota mensual "de catálogo", redondeada a 2 decimales.
    ///
    /// Es lo que se muestra para las cuotas 1..n−1. La última la ajusta `amount(forInstallment:)`
    /// para absorber el residuo.
    var monthlyAmount: Decimal {
        guard totalInstallments > 0 else { return 0 }
        return Self.roundedToCents(totalAmount / Decimal(totalInstallments))
    }

    /// Monto de la cuota `n` (1-indexed), con el residuo absorbido en la última.
    ///
    /// Sin esto, un plan de $1.000.000 en 12 daba doce cuotas de `83333,333…` que la UI mostraba
    /// como $83.333 — y 12 × $83.333 = **$999.996**. Faltaban $4 que no estaban en ninguna cuota,
    /// y el usuario que suma sus cuotas no llega al total de su plan. Cualquier tarjeta factura al
    /// revés: n−1 cuotas parejas y la última cierra la diferencia exacta.
    func amount(forInstallment n: Int) -> Decimal {
        guard totalInstallments > 0, n >= 1, n <= totalInstallments else { return 0 }
        let regular = monthlyAmount
        guard n == totalInstallments else { return regular }
        // La última cierra: total − lo ya facturado. Así la suma da SIEMPRE el total exacto.
        return totalAmount - regular * Decimal(totalInstallments - 1)
    }

    static func roundedToCents(_ value: Decimal) -> Decimal {
        var input = value
        var output = Decimal()
        NSDecimalRound(&output, &input, 2, .plain)
        return output
    }

    /// Fecha efectiva de la cuota número `n` (1-indexed).
    func periodFor(installment n: Int) -> (year: Int, month: Int) {
        let offset = n - 1
        var y = startYear
        var m = startMonth + offset
        while m > 12 { m -= 12; y += 1 }
        while m < 1  { m += 12; y -= 1 }
        return (y, m)
    }

    /// Indica si la cuota del (año, mes) dado corresponde a este plan.
    func installmentNumber(for year: Int, month: Int) -> Int? {
        let startMonths = startYear * 12 + (startMonth - 1)
        let targetMonths = year * 12 + (month - 1)
        let diff = targetMonths - startMonths
        guard diff >= 0, diff < totalInstallments else { return nil }
        return diff + 1
    }

    enum PlanStatus: String, Codable, Hashable, Sendable {
        case active, completed, cancelled
    }
}

/// Ledger mensual de cuota (una fila por mes planificado).
struct InstallmentPayment: Codable, Identifiable, Hashable, Sendable {
    let id: UUID
    var planId: UUID
    var periodYear: Int
    var periodMonth: Int
    var installmentNumber: Int
    var amount: Decimal
    var paid: Bool
    var paidAt: Date?
    var transactionId: UUID?

    enum CodingKeys: String, CodingKey {
        case id
        case planId = "plan_id"
        case periodYear = "period_year"
        case periodMonth = "period_month"
        case installmentNumber = "installment_number"
        case amount
        case paid
        case paidAt = "paid_at"
        case transactionId = "transaction_id"
    }
}

File: metacasa-flutter/lib/models/installment.dart
import 'package:decimal/decimal.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

import '../core/utils/decimal_converter.dart';

part 'installment.freezed.dart';
part 'installment.g.dart';

/// Estado del plan de cuotas. OJO: `cancelled` con doble-L (wire-value real).
enum PlanStatus {
  @JsonValue('active')
  active,
  @JsonValue('completed')
  completed,
  @JsonValue('cancelled')
  cancelled,
}

/// Plan de cuotas (ej. iPhone en 12 cuotas). Espejo 1:1 de
/// `public.installment_plans`.
@freezed
abstract class InstallmentPlan with _$InstallmentPlan {
  const InstallmentPlan._();

  const factory InstallmentPlan({
    required String id,
    @JsonKey(name: 'household_id') required String householdId,
    required String name,
    @JsonKey(name: 'total_amount')
    @DecimalConverter()
    required Decimal totalAmount,
    @JsonKey(name: 'total_installments') required int totalInstallments,
    required String currency,
    @JsonKey(name: 'start_year') required int startYear,
    @JsonKey(name: 'start_month') required int startMonth,
    String? category,
    @JsonKey(name: 'account_id') String? accountId,
    String? note,
    required PlanStatus status,
    @JsonKey(name: 'created_by') required String createdBy,
    @JsonKey(name: 'created_at') DateTime? createdAt,
    @JsonKey(name: 'updated_at') DateTime? updatedAt,
  }) = _InstallmentPlan;

  factory InstallmentPlan.fromJson(Map<String, dynamic> json) =>
      _$InstallmentPlanFromJson(json);

  /// Monto de cada cuota (total / número de cuotas). Matemática en `Decimal`.
  Decimal get monthlyAmount {
    if (totalInstallments <= 0) return Decimal.zero;
    return (totalAmount / Decimal.fromInt(totalInstallments))
        .toDecimal(scaleOnInfinitePrecision: 10);
  }

  /// Año/mes efectivo de la cuota número [n] (1-indexed).
  ({int year, int month}) periodFor(int n) {
    final int offset = n - 1;
    var y = startYear;
    var m = startMonth + offset;
    while (m > 12) {
      m -= 12;
      y += 1;
    }
    while (m < 1) {
      m += 12;
      y -= 1;
    }
    return (year: y, month: m);
  }

  /// Número de cuota (1-indexed) que corresponde al (año, mes) dado, o `null`
  /// si ese período no pertenece al plan.
  int? installmentNumber({required int year, required int month}) {
    final int startMonths = startYear * 12 + (startMonth - 1);
    final int targetMonths = year * 12 + (month - 1);
    final int diff = targetMonths - startMonths;
    if (diff < 0 || diff >= totalInstallments) return null;
    return diff + 1;
  }
}

/// Ledger mensual de cuota (una fila por mes planificado). Espejo 1:1 de
/// `public.installment_payments`.
@freezed
abstract class InstallmentPayment with _$InstallmentPayment {
  const factory InstallmentPayment({
    required String id,
    @JsonKey(name: 'plan_id') required String planId,
    @JsonKey(name: 'period_year') required int periodYear,
    @JsonKey(name: 'period_month') required int periodMonth,
    @JsonKey(name: 'installment_number') required int installmentNumber,
    @DecimalConverter() required Decimal amount,
    required bool paid,
    @JsonKey(name: 'paid_at') DateTime? paidAt,
    @JsonKey(name: 'transaction_id') String? transactionId,
  }) = _InstallmentPayment;

  factory InstallmentPayment.fromJson(Map<String, dynamic> json) =>
      _$InstallmentPaymentFromJson(json);
}
